File: fxgl-base/src/main/java/com/almasb/fxgl/particle/ParticleComponent.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.particle;

import com.almasb.fxgl.core.collection.Array;
import com.almasb.fxgl.core.collection.UnorderedArray;
import com.almasb.fxgl.core.pool.Pools;
import com.almasb.fxgl.entity.Entity;
import com.almasb.fxgl.entity.component.Component;
import com.almasb.fxgl.entity.components.PositionComponent;
import com.almasb.fxgl.util.EmptyRunnable;

import java.util.Iterator;

import static com.almasb.fxgl.util.BackportKt.forEach;

/**
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public class ParticleComponent extends Component {

    private ParticleEmitter emitter;
    private Entity parent = new Entity();

    protected Array<Particle> particles = new UnorderedArray<>(256);

    private PositionComponent position;

    private Runnable onFinished = EmptyRunnable.INSTANCE;

    /**
     * Constructs particle component with specified emitter.
     *
     * @param emitter particle emitter
     */
    public ParticleComponent(ParticleEmitter emitter) {
        this.emitter = emitter;
    }

    protected ParticleComponent() {}

    @Override
    public void onUpdate(double tpf) {
        if (emitter == null)
            return;

        if (parent.getWorld() == null) {
            entity.getWorld().addEntity(parent);
        }

        particles.addAll(emitter.emit(position.getX(), position.getY()));

        for (Iterator<Particle> it = particles.iterator(); it.hasNext(); ) {
            Particle p = it.next();
            if (p.update(tpf)) {
                it.remove();

                parent.getView().removeNode(p.getView());
                Pools.free(p);
            } else {
                if (p.getView().getParent() == null)
                    parent.getView().addNode(p.getView());
            }
        }

        if (particles.isEmpty() && emitter.isFinished()) {
            onFinished.run();
        }
    }

    @Override
    public void onRemoved() {
        forEach(particles, Pools::free);
        particles.clear();

        parent.removeFromWorld();
    }

    /**
     * @return particle emitter attached to component
     */
    public final ParticleEmitter getEmitter() {
        return emitter;
    }

    public final void setOnFinished(Runnable onFinished) {
        this.onFinished = onFinished;
    }
}

File: fxgl-effects/src/main/kotlin/com/almasb/fxgl/particle/ParticleComponent.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.particle

import com.almasb.fxgl.core.collection.UnorderedArray
import com.almasb.fxgl.core.pool.Pools
import com.almasb.fxgl.core.util.EmptyRunnable
import com.almasb.fxgl.entity.Entity
import com.almasb.fxgl.entity.component.Component

/**
 * Allows adding particle effects to an entity.
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
open class ParticleComponent(val emitter: ParticleEmitter) : Component() {

    var onFinished: Runnable = EmptyRunnable

    /**
     * This is the entity whose view is used to render particles.
     * Use of extra entity allows to render particles independently from
     * the entity to which this component is attached. Otherwise, the entire
     * set of emitted particles will be moved based on entity's view.
     */
    private val parent = Entity()

    private val particles = UnorderedArray<Particle>(256)

    override fun onUpdate(tpf: Double) {
        if (parent.world == null) {
            entity.world.addEntity(parent)
        }

        particles.addAll(emitter.emit(entity.x, entity.y))

        val iter = particles.iterator()
        while (iter.hasNext()) {
            val p = iter.next()

            if (p.update(tpf)) {
                iter.remove()

                parent.viewComponent.removeChild(p.view)
                Pools.free(p)
            } else {
                if (p.view.parent == null)
                    parent.viewComponent.addChild(p.view)
            }
        }

        if (particles.isEmpty && emitter.isFinished) {
            onFinished.run();
        }
    }

    override fun onRemoved() {
        particles.forEach { Pools.free(it) }
        particles.clear()

        parent.removeFromWorld()
    }
}